src/shared/libcrypt-util.c

```diff
@@ -74,6 +90,60 @@ int make_salt(char **ret) {
 #endif
 }
 
+#if !HAVE_CRYPT_RA
+/* Provide a poor man's fallback that uses a fixed size buffer. */
+
+static char* systemd_crypt_ra(const char *phrase, const char *setting, void **data, int *size) {
+        assert(data);
+        assert(size);
+
+        /* We allocate the buffer because crypt(3) says: struct crypt_data may be quite large (32kB in this
+         * implementation of libcrypt; over 128kB in some other implementations). This is large enough that
+         * it may be unwise to allocate it on the stack. */
+
+        if (!*data) {
+                *data = new0(struct crypt_data, 1);
+                if (!*data) {
+                        errno = -ENOMEM;
+                        return NULL;
+                }
+
+                *size = (int) (sizeof(struct crypt_data));
+        }
+
+        return crypt_r(phrase, setting, *data);
```

Comment:
From crypt(3),
```
     Upon error, crypt_rn and crypt_ra return a null pointer.  crypt_r and
     crypt may also return a null pointer, or they may return a pointer to the
     invalid hash, depending on how libcrypt was configured.  (The option to
     return the invalid hash is for compatibility with old applications that
     assume that crypt cannot return a null pointer.  See PORTABILITY NOTES
     below.)
```
So, maybe, we need to check that the return value rather than simply return it.

Reply:
Also, `systemd_crypt_ra()` updates `*data` even if `crypt_r()` fails. We usually set return value only when the function succeeds.
Of course, in all our use cases, `_cleanup_` attribute is set, so it is harmless.

Reply:
> we need to check that the return value rather than simply return it.

OK, I added a test now.

> Also, `systemd_crypt_ra()` updates `*data` even if `crypt_r()`

I think that's OK. That's what crypt_ra() does, and we just mimic that in the replacement function.

Reply:
actually, it doesn't update it, if you look closely. That's because `*data` is only updated if it is NULL. And if it then fails, we just reassign `NULL` to it, thus no actual change. Or in other words: yes we update the parameter on failure, but we update it to the very same value it already had, which should be OK.

Reply:
Veeeery strictly speaking, it does update it on error. The call to crypt_r can also fail. But I think that's fine.